src/ark/utils/data_utils.py

```diff
@@ -54,119 +57,199 @@ def save_fov_mask(fov, data_dir, mask_data, sub_dir=None, name_suffix=''):
-    relabeled_img = np.vectorize(
-        lambda x: labels_dict.get(x, default_label) if x != 0 else 0
-    )(img).astype('int16')
-
-    return relabeled_img
+    Args:
+        data (pd.DataFrame):
+            A cell table with the cell label column and the cluster column.
+        fov_col (str):
+            The name of the column in the cell table that contains the FOV ID.
+        label_column (str):
+            The name of the column in the cell table that contains the cell label.
+        cluster_column (str):
+            The name of the column in the cell table that contains the cluster label.
 
 
-def label_cells_by_cluster(fov, all_data, label_map, fov_col=settings.FOV_ID,
-                           cell_label_column=settings.CELL_LABEL,
-                           cluster_column=settings.KMEANS_CLUSTER):
-    """Translates cell-ID labeled images according to the clustering assignment.
+    Fields:
+        unique_fovs (List[str]):
+            A list of all unique FOVs in the cell table.
+        unassigned_id (int):
+            The cluster ID that is assigned to cells that are not assigned to a cluster.
+        mapping (pd.core.groupby.DataFrameGroupBy):
+            A GroupBy object that contains the mapping from the cell label to the cluster label for
+            each FOV, sorted by Segmentation label.
+    """
+    data: pd.DataFrame
+    fov_col: str
+    label_column: str
+    cluster_column: str
+
+    # Initialized in __post_init__
+    unique_fovs: List[str] = field(init=False)
+    unassigned_id: int = field(init=False)
+    mapping: DataFrameGroupBy = field(init=False)
+
+    def __post_init__(self) -> None:
+
+        self.data[self.label_column] = self.data[self.label_column].astype(np.int64)
+        self.data[self.cluster_column] = self.data[self.cluster_column].astype(np.int64)
+
+        self.unique_fovs = ns.natsorted(self.data[self.fov_col].unique().tolist())
+
+        self.unassigned_id: int = self.data[self.cluster_column].max() + 1
+
+        self.mapping: DataFrameGroupBy = (
+            pd.concat(
+                [
+                    self.data[[self.fov_col, self.label_column, self.cluster_column]],
+                    # Make sure that the cluster 0 maps to the metacluster 0
+                    pd.DataFrame(
+                        data={
+                            self.fov_col: self.unique_fovs,
+                            self.label_column: np.repeat(
+                                0, repeats=len(self.unique_fovs)
+                            ),
+                            self.cluster_column: np.repeat(
+                                0, repeats=len(self.unique_fovs)
+                            ),
+                        }
+                    ),
+                ]
+            )
+            # sort by FOV first, then by segmentation label
+            # Then group by FOV
+            .sort_values(by=[self.fov_col, self.label_column])
```

Comment:
So in the end, this dataclass stores every FOV, and for each FOV, the mapping between each individual segmentation ID and the corresponding cluster ID, is that correct?

Have you tried running this function, and using this dataclass, on a very large cell table? My concern is that doing all the mapping ahead of time and storing the results, rather than in real time when needed, is going to cause this be slow

Reply:
Optimized this and ran ran some performance metrics with [scalene](https://github.com/plasma-umass/scalene):
1. Switched from a `dataclass` to a regular `class`, therefore we will only be extracting the `fov_column` `label_column` (`segmentation_label`), and the `cluster_column` from the cell table, whereas the`dataclass` would contain both the original DataFrame and the 
2. Benchmarked the new process with a Cell Table consisting of 10 Million cells, 2000 FOVs, and 40 MetaClusters totaling 7.9Gb. Creating the mapping ahead of time requires additional 1.4 GB of memory usage for the whole table. This step took less than 3 seconds.